File: checks/python/src/section508/checks/no_change_on_input.py
import re

from section508.dom import attr, select

NAVIGATE = re.compile(
    r"location\s*=|window\.open|document\.location|href\s*=|this\.submit|form\.submit",
    re.I,
)


def check(root):
    if not root:
        return []

    for control in select(root, "select, input, textarea, [onchange], [oninput]"):
        handler = f"{attr(control, 'onchange') or ''} {attr(control, 'oninput') or ''}"
        if NAVIGATE.search(handler):
            return ["Changing this value would change context."]
    return []

File: checks/ruby/lib/section508/checks/no_change_on_input.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Changing a value must not navigate unless the person asks.
    class NoChangeOnInput
      NAVIGATE = /location\s*=|window\.open|document\.location|href\s*=|this\.submit|form\.submit/i

      def self.call(root)
        return [] unless root&.respond_to?(:css)

        Dom.elements(root, "select, input, textarea, [onchange], [oninput]").each do |control|
          handler = "#{control["onchange"]} #{control["oninput"]}"
          return ["Changing this value would change context."] if NAVIGATE.match?(handler)
        end
        []
      end
    end
  end
end
